group-coordinator/src/test/java/org/apache/kafka/coordinator/group/streams/topics/RepartitionTopicsTest.java

```diff
@@ -0,0 +1,273 @@
+    public void shouldThrowStreamsMissingSourceTopicsExceptionIfMissingSourceTopics() {
+        List<Subtopology> subtopologyToSubtopology = List.of(SUBTOPOLOGY1, SUBTOPOLOGY2);
+        Function<String, OptionalInt> topicPartitionCountProvider =
+            s -> Objects.equals(s, SOURCE_TOPIC_NAME1) ? OptionalInt.empty() :
+                (TOPICS.contains(s) ? OptionalInt.of(3) : OptionalInt.empty());
+        final RepartitionTopics repartitionTopics = new RepartitionTopics(
+            LOG_CONTEXT,
+            subtopologyToSubtopology,
+            topicPartitionCountProvider
+        );
+
+        final TopicConfigurationException exception = assertThrows(TopicConfigurationException.class,
+            repartitionTopics::setup);
+
+        assertNotNull(exception);
+        assertEquals(Status.MISSING_SOURCE_TOPICS, exception.status());
+        assertEquals("Missing source topics: source1", exception.getMessage());
+    }
+
+    @Test
+    public void shouldThrowStreamsMissingSourceTopicsExceptionIfPartitionCountCannotBeComputedForAllRepartitionTopics() {
+        List<Subtopology> subtopologyToSubtopology = List.of(SUBTOPOLOGY1, SUBTOPOLOGY_WITHOUT_PARTITION_COUNT);
+        Function<String, OptionalInt> topicPartitionCountProvider = s -> TOPICS.contains(s) ? OptionalInt.of(3) : OptionalInt.empty();
+        final RepartitionTopics repartitionTopics = new RepartitionTopics(
+            LOG_CONTEXT,
+            subtopologyToSubtopology,
+            topicPartitionCountProvider
+        );
+
+        TopicConfigurationException exception = assertThrows(TopicConfigurationException.class, repartitionTopics::setup);
+
+        assertEquals(Status.MISSING_SOURCE_TOPICS, exception.status());
+        assertEquals(
+            "Failed to compute number of partitions for all repartition topics, make sure all user input topics are created "
+                + "and all pattern subscriptions match at least one topic in the cluster",
+            exception.getMessage()
+        );
+    }
+
+    @Test
+    public void shouldSetRepartitionTopicPartitionCountFromUpstreamExternalSourceTopic() {
+        final Subtopology subtopology = new Subtopology()
+            .setSubtopologyId("SUBTOPOLOGY0")
+            .setSourceTopics(List.of(SOURCE_TOPIC_NAME1, REPARTITION_TOPIC_NAME2))
+            .setRepartitionSinkTopics(List.of(REPARTITION_TOPIC_NAME1, REPARTITION_TOPIC_WITHOUT_PARTITION_COUNT))
+            .setRepartitionSourceTopics(List.of(
+                REPARTITION_TOPIC_INFO1,
+                REPARTITION_TOPIC_INFO2,
+                REPARTITION_TOPIC_INFO_WITHOUT_PARTITION_COUNT
+            ))
+            .setStateChangelogTopics(Collections.emptyList());
+        List<Subtopology> subtopologyToSubtopology = List.of(
+            subtopology,
+            SUBTOPOLOGY_WITHOUT_PARTITION_COUNT
+        );
+        Function<String, OptionalInt> topicPartitionCountProvider = s -> TOPICS.contains(s) ? OptionalInt.of(3) : OptionalInt.empty();
```

Comment:
```suggestion
        int expectedPartitionCount = 3
        Function<String, OptionalInt> topicPartitionCountProvider = s -> TOPICS.contains(s) ? OptionalInt.of(expectedPartitionCount) : OptionalInt.empty();
```
and use `expectedPartitionCount` in the verification on line 187.

Reply:
Hmm - defined the source topics globally now, so I think it should be fine to use a constant three here (what you proposed doesn't work easily anymore).